File: powermanager/src/main/java/com/pyamsoft/powermanager/service/ForegroundService.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.powermanager.service;

import android.content.Context;
import android.content.Intent;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationManagerCompat;
import com.pyamsoft.powermanager.Injector;
import com.pyamsoft.powermanager.receiver.ScreenOnOffReceiver;
import com.pyamsoft.pydroid.ui.app.AutoRestartService;
import javax.inject.Inject;
import timber.log.Timber;

public class ForegroundService extends AutoRestartService {

  static final int NOTIFICATION_ID = 1000;
  @NonNull private static final String EXTRA_RESTART_TRIGGERS = "EXTRA_RESTART_TRIGGERS";
  @Inject ForegroundPresenter presenter;
  private ScreenOnOffReceiver screenOnOffReceiver;
  private NotificationManagerCompat notificationManager;

  /**
   * Force the service On
   */
  public static void start(@NonNull Context context) {
    context.getApplicationContext()
        .startService(new Intent(context.getApplicationContext(), ForegroundService.class));
  }

  /**
   * Force the service Off
   */
  public static void stop(@NonNull Context context) {
    context.getApplicationContext()
        .stopService(new Intent(context.getApplicationContext(), ForegroundService.class));
  }

  /**
   * Restart the triggers
   */
  public static void restartTriggers(@NonNull Context context) {
    final Context appContext = context.getApplicationContext();
    final Intent service = new Intent(appContext, ForegroundService.class);
    service.putExtra(EXTRA_RESTART_TRIGGERS, true);
    Timber.d("Restart Power Triggers");
    appContext.startService(service);
  }

  @Nullable @Override public IBinder onBind(Intent intent) {
    return null;
  }

  @Override public void onCreate() {
    super.onCreate();
    Injector.get().provideComponent().inject(this);

    notificationManager = NotificationManagerCompat.from(getApplicationContext());
    notificationManager.cancel(NOTIFICATION_ID);

    presenter.setForegroundState(true);
    presenter.queueRepeatingTriggerJob();

    screenOnOffReceiver = new ScreenOnOffReceiver(this);
    screenOnOffReceiver.register();
  }

  @Override public void onDestroy() {
    super.onDestroy();
    presenter.setForegroundState(false);
    screenOnOffReceiver.unregister();

    presenter.stop();
    presenter.destroy();

    stopForeground(true);

    notificationManager.notify(NOTIFICATION_ID, presenter.hangNotification());
  }

  @Override public int onStartCommand(Intent intent, int flags, int startId) {
    if (intent != null) {
      if (intent.getBooleanExtra(EXTRA_RESTART_TRIGGERS, false)) {
        presenter.restartTriggerAlarm();
      }
    }
    presenter.startNotification(notification -> startForeground(NOTIFICATION_ID, notification));
    return START_STICKY;
  }
}

File: powermanager/src/main/java/com/pyamsoft/powermanager/service/ForegroundService.kt
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.powermanager.service

import android.app.Notification
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.support.v4.app.NotificationManagerCompat
import com.pyamsoft.powermanager.Injector
import com.pyamsoft.powermanager.receiver.ScreenOnOffReceiver
import com.pyamsoft.powermanager.service.ForegroundPresenter.NotificationCallback
import com.pyamsoft.pydroid.ui.app.AutoRestartService
import timber.log.Timber
import javax.inject.Inject

class ForegroundService : AutoRestartService() {
  @field:Inject lateinit internal var presenter: ForegroundPresenter
  private lateinit var screenOnOffReceiver: ScreenOnOffReceiver
  private lateinit var notificationManager: NotificationManagerCompat

  override fun onBind(intent: Intent): IBinder? {
    return null
  }

  override fun onCreate() {
    super.onCreate()
    Injector.get().provideComponent().inject(this)

    notificationManager = NotificationManagerCompat.from(applicationContext)
    notificationManager.cancel(NOTIFICATION_ID)

    presenter.setForegroundState(true)
    presenter.queueRepeatingTriggerJob()

    screenOnOffReceiver = ScreenOnOffReceiver(this)
    screenOnOffReceiver.register()
  }

  override fun onDestroy() {
    super.onDestroy()
    presenter.setForegroundState(false)
    screenOnOffReceiver.unregister()

    presenter.stop()
    presenter.destroy()

    stopForeground(true)

    notificationManager.notify(NOTIFICATION_ID, presenter.hangNotification())
  }

  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
    if (intent != null) {
      if (intent.getBooleanExtra(EXTRA_RESTART_TRIGGERS, false)) {
        presenter.restartTriggerAlarm()
      }
    }
    presenter.startNotification(object : NotificationCallback {
      override fun onStartNotificationInForeground(notification: Notification) {
        startForeground(NOTIFICATION_ID, notification)
      }
    })
    return Service.START_STICKY
  }

  companion object {
    const internal val NOTIFICATION_ID = 1000
    private const val EXTRA_RESTART_TRIGGERS = "EXTRA_RESTART_TRIGGERS"

    /**
     * Force the service On
     */
    @JvmStatic fun start(context: Context) {
      context.applicationContext.startService(
          Intent(context.applicationContext, ForegroundService::class.java))
    }

    /**
     * Force the service Off
     */
    @JvmStatic fun stop(context: Context) {
      context.applicationContext.stopService(
          Intent(context.applicationContext, ForegroundService::class.java))
    }

    /**
     * Restart the triggers
     */
    @JvmStatic fun restartTriggers(context: Context) {
      val appContext = context.applicationContext
      val service = Intent(appContext, ForegroundService::class.java)
      service.putExtra(EXTRA_RESTART_TRIGGERS, true)
      Timber.d("Restart Power Triggers")
      appContext.startService(service)
    }
  }
}
